src/eos_gates.js

```diff
@@ -143,6 +245,162 @@ const Application = new Lang.Class({
     },
 });
 
+function spawnProcess(argv=[]) {
+    return Gio.Subprocess.new(argv, Gio.SubprocessFlags.NONE);
+}
+
+function getAppStoreAppId(remote, appId) {
+    let installation = Flatpak.Installation.new_system(null);
+    let flatpakRemote = null;
+
+    try {
+        flatpakRemote = installation.get_remote_by_name(remote, null);
+    } catch (e) {
+        logError(e, 'Could not find flatpak remote %s: %s'.format(remote));
+    }
+
+    // Get the default branch now to construct the full unique ID GNOME Software expects.
+    if (flatpakRemote) {
+        let defaultBranch = flatpakRemote.get_default_branch()
+        if (defaultBranch)
+            return 'system/flatpak/%s/desktop/%s.desktop/%s'.format(remote,
+                                                                    appId,
+                                                                    defaultBranch);
+    }
+
+    return appId;
+}
+
+// Happens when we launch gnome-software for a known flatpak replacement
+// of an unsupported binary.
+const EVENT_LAUNCHED_INSTALLER_FOR_FLATPAK = 'e98bf6d9-8511-44f9-a1bd-a1d0518934b9';
+
+// Happens when we launch a known flatpak replacement of an unsupported binary
+// directly.
+const EVENT_LAUNCHED_EXISTING_FLATPAK = '192f39dd-79b3-4497-99fa-9d8aea28760c';
+
+// Happens when we launch gnome-software to install an 'equivalent' app for
+// an unsupported binary.
+const EVENT_LAUNCHED_EQUIVALENT_INSTALLER_FOR_FLATPAK = '7de69d43-5f6b-4bef-b5f3-a21295b79185';
+
+// Happens when we launch an 'equivalent' app for an unsupported binary directly. 
+const EVENT_LAUNCHED_EQUIVALENT_EXISTING_FLATPAK = '00d7bc1e-ec93-4c53-ae78-a6b40450be4a';
+
+function installAppFromStore(replacement, originalPayload) {
+    let appStoreId = getAppStoreAppId(replacement.flatpakInfo.remote,
+                                      replacement.flatpakInfo.id);
+    recordMetrics(replacement.replacementInfo ?
+                  EVENT_LAUNCHED_EQUIVALENT_INSTALLER_FOR_FLATPAK :
+                  EVENT_LAUNCHED_INSTALLER_FOR_FLATPAK,
+                  new GLib.Variant('(sas)', [replacement.flatpakInfo.id, originalPayload]));
+    Gio.DBusActionGroup.get(Gio.Application.get_default().get_dbus_connection(),
+                            'org.gnome.Software',
+                            '/org/gnome/Software')
+                       .activate_action('details',
+                                        new GLib.Variant('(ss)', [appStoreId, '']));
+}
+
+function launchFlatpakAppForInstallation(installation, replacement) {
+    let ref = null;
+
+    try {
+        ref = installation.get_current_installed_app(replacement.flatpakInfo.id, null);
+    } catch (e) {
+        return false;
+    }
+
+    installation.launch(replacement.flatpakInfo.id,
+                        null,
+                        'master',
+                        ref.get_latest_commit(),
```

Comment:
> does it not fail if you don't quit the application right after? If so, the problem may be that an IPC call does not get to the dbus daemon broker before the mainloop quits

I think this is the most likely answer (which may be a bug in upstream flatpak, since it should really take a callback to launch the application). I'll check if adding a ~2s delay fixes the problem . If so, I think I'd still be hesitant about doing it this way, considering that a system under heavy load may not make the IPC call before we quit the mainloop. Also, it does strike me as odd that quitting the mainloop doesn't run all pending sources first, which one would think the `installation-launch` dbus call would be one of them....